cluster-provision/gocli/cmd/run.go

```diff
@@ -586,212 +605,178 @@ func run(cmd *cobra.Command, args []string) (retErr error) {
+			return fmt.Errorf("Cleaning up scripts dir failed: %s", err)
 		}
+	}
 
-		if enableAudit {
-			ok, err := docker.Exec(cli, nodeContainer(prefix, nodeName),
-				[]string{"/bin/bash", "-c", "ssh.sh touch /home/vagrant/enable_audit"}, os.Stdout)
-			if err != nil {
-				return err
-			}
+	// If background flag was specified, we don't want to clean up if we reach that state
+	if !background {
+		wg.Wait()
+		stop <- fmt.Errorf("Done. please clean up")
+	}
 
-			if !ok {
-				return fmt.Errorf("provisioning node %s failed (setting enableAudit phase)", nodeName)
-			}
-		}
+	return nil
+}
 
-		if psaEnabled {
-			success, err := docker.Exec(cli, nodeContainer(prefix, nodeName), []string{"/bin/bash", "-c", "ssh.sh sudo /bin/bash < /scripts/psa.sh"}, os.Stdout)
-			if err != nil {
-				return err
-			}
+func provisionK8sOptions(sshClient sshutils.SSHClient, n *nodesconfig.NodeK8sConfig) error {
+	if n.Ceph {
+		if err := sshClient.SSH("/scripts/rook-ceph.sh"); err != nil {
+			return fmt.Errorf("provisioning Ceph CSI failed: %s", err)
+		}
+	}
 
-			if !success {
-				return fmt.Errorf("provisioning node %s failed", nodeName)
-			}
+	if n.NfsCsi {
+		if err := sshClient.SSH("/scripts/nfs-csi.sh"); err != nil {
+			return fmt.Errorf("provisioning NFS CSI failed: %s", err)
 		}
+	}
 
-		if success {
-			success, err = docker.Exec(cli, nodeContainer(prefix, nodeName), []string{"/bin/bash", "-c", fmt.Sprintf("ssh.sh sudo /bin/bash < /scripts/%s.sh", nodeName)}, os.Stdout)
-		} else {
-			if gpuAddress != "" {
-				// move the assigned PCI device to a vfio-pci driver to prepare for assignment
-				err = prepareDeviceForAssignment(cli, nodeContainer(prefix, nodeName), "", gpuAddress)
-				if err != nil {
-					return err
-				}
-			}
-			success, err = docker.Exec(cli, nodeContainer(prefix, nodeName), []string{"/bin/bash", "-c", "ssh.sh sudo /bin/bash < /scripts/nodes.sh"}, os.Stdout)
+	if n.Istio {
+		if err := sshClient.SSH("/scripts/istio.sh"); err != nil {
+			return fmt.Errorf("deploying Istio service mesh failed: %s", err)
 		}
+	}
 
-		if err != nil {
-			return err
+	if n.Prometheus {
+		var params string
+		if n.Alertmanager {
+			params += "--alertmanager true "
 		}
 
-		if !success {
-			return fmt.Errorf("provisioning node %s failed", nodeName)
+		if n.Grafana {
+			params += "--grafana true "
 		}
 
-		go func(id string) {
-			cli.ContainerWait(ctx, id, container.WaitConditionNotRunning)
-			wg.Done()
-		}(node.ID)
+		if err := sshClient.SSH(fmt.Sprintf("-s -- %s < /scripts/prometheus.sh", params)); err != nil {
+			return fmt.Errorf("deploying Prometheus operator failed: %s", err)
+		}
 	}
+	return nil
+}
 
-	if cephEnabled {
-		nodeName := nodeNameFromIndex(1)
-		success, err := docker.Exec(cli, nodeContainer(prefix, nodeName), []string{
-			"/bin/bash",
-			"-c",
-			"ssh.sh sudo /bin/bash < /scripts/rook-ceph.sh",
-		}, os.Stdout)
+func provisionNode(sshClient sshutils.SSHClient, n *nodesconfig.NodeLinuxConfig) error {
```

Comment:
Sure, In my opinion the config is just a config. you should declare a method on a struct when the struct is whats doing the work, for example `client.Apply` or `docker.Exec` as it shows to who's reading the code the relation between the two.

In subsequent PR's i will be introducing a `KubevirtProvider` type, which can implement this method and it will maintain this relationship i am talking about

if you think otherwise i'd love to hear your take

Reply:
You could i.e. call the func `Apply` if it's a receiver method. This would then read

```
myNodeConfig.Apply(sshClient ...)
```

but yeah, we can leave that for later.